### ISSUE TITLE
Exchange Rate Difference Incorrectly Stored as Unallocated Amount

### ISSUE THREAD

**Issue by Manaa0-0** [frappe/erpnext#54947]
## Payment Flow Analysis

| Document | Currency | Exchange Rate | Amount (Original) | Amount (Base - YER) |
|----------|----------|---------------|-------------------|---------------------|
| **Sales Invoice** | USD | 531 | 110 USD | 58,410 YER |
| **Payment Entry** | USD | 600 | 66,000 USD (paid) / 110 USD (received) | 66,000 YER (paid) / 110 YER (received) |

## Key Observations from Your Data:

1. **Invoice**: 110 USD @ rate 531 = **58,410 YER** (fully paid)
2. **Payment Entry**:
   - `paid_amount`: 66,000 USD
   - `received_amount`: 110 USD  
   - `source_exchange_rate`: 1
   - `target_exchange_rate`: 600
   - `total_allocated_amount`: 58,410 YER (matches invoice)
   - `unallocated_amount`: 7,590 YER

## The Problem:

When a Payment Entry is created where:
- **Party/Customer currency**: YER (company's receivable account is in YER)
- **Invoice currency**: USD 
- **Payment currency**: USD

The system should recognize that:
- Invoice is for 58,410 YER (110 USD × 531)
- Payment allocates 58,410 YER to the invoice
- But the payment's base calculation shows 66,000 YER vs 58,410 YER needed
- The difference (7,590 YER) should create an **exchange gain/loss journal entry** or be properly handled

Instead, ERPNext is registering the difference as **unallocated_amount** rather than recognizing it as exchange difference.

## Root Cause:

This is a known limitation in ERPNext's multi-currency payment reconciliation when:
1. The party's account currency differs from the invoice currency
2. The payment's exchange rate differs from the invoice's exchange rate
3. ERPNext doesn't automatically create the exchange gain/loss journal entry for this scenario

The payment gets partially allocated (58,410 YER matches the invoice), but the remaining 7,590 YER sits as unallocated because the system doesn't recognize it as an exchange difference that needs to be booked.

**Environment**
- ERPNext Version: v16.12.0
- Frappe Version: v16.10.0
- Deployment: Self-hosted

**Comment by Asobu01** [comment]
I can pick up frappe/erpnext#54947 now.
Given "Exchange Rate Difference Incorrectly Stored as Unallocated Amount", I will begin with reproducing the reported issue first, then applying a focused fix with clear validation.
I can post a reproducible checklist/log snippet in ~30 minutes so you can verify execution speed before the full PR.
Plan for this thread: repro + root-cause summary -> focused patch (plus tests/docs if needed) -> review-ready PR with validation notes.
Expected first PR: ~8h. @Asobu01

**Comment by Asobu01** [comment]
Quick follow-up on `frappe/erpnext#54947`. I am still ready to start immediately and can post the first repro/checklist note before the draft PR if helpful. If you want me on this one, I can begin now. (@Asobu01)

**Comment by Manaa0-0** [comment]
@DipenFrappe  please could you see this issue thank for your help

**Comment by Asobu01** [comment]
Thanks for the clarification on "@DipenFrappe please could you see this issue thank for your help".
I will align with these requirements and post a concrete checkpoint in ~20 minutes (repro/logs or patch summary), then submit a draft PR in ~8h.

**Comment by DipenFrappe** [comment]
> 1. The party's account currency differs from the invoice currency

Base/Company Currency is USD, right?
Party currency YER, right?
Invoice transaction currency is YER, right?
Bank Account is in which currency?

The system books exchange gain or loss automatically when a payment is created.

**Comment by Manaa0-0** [comment]
@DipenFrappe

This issue occurs after enabling multi-currency transactions against a single customer account.

The expected workflow in a modern ERP is that we should not need to create separate customer accounts such as:

- Customer Name - USD
- Customer Name - YER
- Customer Name - EUR

Scenario

- Company/Base Currency: YER
- Party Account Currency: YER
- Sales Invoice Currency: USD
- Payment Entry Currency: USD
- Bank Account Currency: USD

Example

Sales Invoice

- Invoice Amount: 110 USD
- Exchange Rate: 531
- Base Amount: 58,410 YER

Payment Entry

- Payment Amount: 110 USD
- Payment Exchange Rate: 600
- Base Amount Paid: 66,000 YER

ERPNext correctly allocates:

- 58,410 YER against the invoice

However, the remaining:

- 7,590 YER

is treated as "unallocated_amount".

Expected Behavior

The difference between:

- Invoice base amount (58,410 YER)
  and
- Payment base amount (66,000 YER)

should be recognized as an exchange gain/loss and automatically booked through Exchange Gain/Loss accounting.

The payment should still be considered fully allocated because the customer paid the full 110 USD invoice amount.

**Current Behavior**

When the payment exchange rate is **higher** than the invoice exchange rate:

- ERPNext leaves the difference as "**unallocated_amount**"

When the payment exchange rate is **lower** than the invoice exchange rate:

- **ERPNext throws an allocation error**

Root Issue

ERPNext currently struggles in this scenario when:

- Party account currency = Base currency (YER)
- Transaction currency = Foreign currency (USD)
- Payment exchange rate differs from invoice exchange rate

The system allocates based on base currency values instead of recognizing that the foreign currency obligation (110 USD) has already been fully settled.

This appears to be a limitation in multi-currency reconciliation logic when using a single receivable account for multiple currencies.

**Comment by Asobu01** [comment]
Thanks for the clarification on "@DipenFrappe This issue occurs after enabling multi-currency transactions against a single cu...".
I will align with these requirements and post a concrete checkpoint in ~20 minutes (repro/logs or patch summary), then submit a draft PR in ~8h.

**Comment by Manaa0-0** [comment]
@DipenFrappe you can show this code to frappe developer it solve the problem and make erpnext act like modern ERP systems in sell invoice and purchase invoice.
	

---

Complete, Final, Strongest Code

```python
class PaymentEntry(AccountsController):
    # ... existing code ...
    
    # Only SI/PI have guaranteed conversion_rate semantics compatible with
    # this allocation logic. Other doctypes may use different rate conventions
    # or lack stable invoice-level FX metadata.
    SUPPORTED_FX_DOCTYPES = ("Sales Invoice", "Purchase Invoice")

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        # Initialize cached FX scenario attributes deterministically
        self._payment_rate_allocation_result = None
        self._payment_rate_allocation_invoice_map = {}

    def _is_payment_rate_allocation(self):
        """
        Detect scenario where allocated_amount is in payment foreign currency (FC)
        but the existing code incorrectly uses party-account rate (1) instead of
        payment-account rate for base currency conversion.
        
        Hard guards (ALL must pass):
        1. References exist and payment is not Internal Transfer
        2. Party account is in company base currency
        3. Payment account is in foreign currency (not base)
        4. All references are supported doctypes (SI/PI)
        5. All references share exactly ONE doctype (no mixed SI/PI)
        6. All references share exactly one currency
        7. Reference currency matches payment account currency
        8. Payment account exists and is valid
        """
        if self._payment_rate_allocation_result is not None:
            return self._payment_rate_allocation_result
        
        if not self.references or self.payment_type == "Internal Transfer":
            self._payment_rate_allocation_result = False
            return False

        if self.payment_type == "Receive":
            payment_account = self.paid_to
            payment_currency = self.paid_to_account_currency
            party_currency = self.paid_from_account_currency
        else:
            payment_account = self.paid_from
            payment_currency = self.paid_from_account_currency
            party_currency = self.paid_to_account_currency

        company_currency = erpnext.get_company_currency(self.company)

        if party_currency != company_currency or payment_currency == company_currency:
            self._payment_rate_allocation_result = False
            return False
        
        allocated_refs = [ref for ref in self.references if flt(ref.allocated_amount)]
        if not allocated_refs:
            self._payment_rate_allocation_result = False
            return False
        
        ref_doctypes = {ref.reference_doctype for ref in allocated_refs}
        if not ref_doctypes.issubset(set(self.SUPPORTED_FX_DOCTYPES)) or len(ref_doctypes) != 1:
            self._payment_rate_allocation_result = False
            return False
        
        ref_doctype = ref_doctypes.pop()
        
        invoice_names = [ref.reference_name for ref in allocated_refs]
        invoices = frappe.get_all(
            ref_doctype,
            filters={"name": ["in", invoice_names]},
            fields=["name", "currency", "conversion_rate"],
        )
        invoice_map = {d.name: d for d in invoices}

        ref_currencies = set()
        for ref in allocated_refs:
            inv = invoice_map.get(ref.reference_name)
            if not inv or not inv.currency:
                self._payment_rate_allocation_result = False
                return False
            ref_currencies.add(inv.currency)

        if len(ref_currencies) != 1 or payment_currency not in ref_currencies:
            self._payment_rate_allocation_result = False
            return False
        
        if not payment_account or not frappe.db.exists("Account", payment_account):
            self._payment_rate_allocation_result = False
            return False

        self._payment_rate_allocation_invoice_map = invoice_map
        self._payment_rate_allocation_result = True
        return True

    def _get_payment_rate_for_base_allocation(self):
        if self.payment_type == "Receive":
            return self.target_exchange_rate
        elif self.payment_type == "Pay":
            return self.source_exchange_rate
        return 1

    def calculate_base_allocated_amount_for_reference(self, d) -> float:
        base_allocated_amount = 0
        if d.reference_doctype in get_advance_payment_doctypes():
            exchange_rate = 1
            if self.payment_type == "Receive":
                exchange_rate = self.source_exchange_rate
            elif self.payment_type == "Pay":
                exchange_rate = self.target_exchange_rate

            base_allocated_amount += flt(
                flt(d.allocated_amount) * flt(exchange_rate), self.precision("base_paid_amount")
            )
        else:
            is_pra = self._is_payment_rate_allocation()
            if is_pra:
                exchange_rate = self._get_payment_rate_for_base_allocation()
            else:
                exchange_rate = 1
                if self.payment_type == "Receive":
                    exchange_rate = self.source_exchange_rate
                elif self.payment_type == "Pay":
                    exchange_rate = self.target_exchange_rate

            base_allocated_amount += flt(
                flt(d.allocated_amount) * flt(exchange_rate), self.precision("base_paid_amount")
            )

            if is_pra and d.reference_doctype in self.SUPPORTED_FX_DOCTYPES:
                inv = self._payment_rate_allocation_invoice_map.get(d.reference_name)
                inv_rate = inv.conversion_rate if inv else frappe.db.get_value(
                    d.reference_doctype, d.reference_name, "conversion_rate", cache=True
                )
                allocated_amount_in_ref_exchange_rate = flt(
                    flt(d.allocated_amount) * flt(inv_rate or 1),
                    self.precision("base_paid_amount")
                )
            else:
                if d.exchange_rate is None:
                    d.exchange_rate = 1
                allocated_amount_in_ref_exchange_rate = flt(
                    flt(d.allocated_amount) * flt(d.exchange_rate),
                    self.precision("base_paid_amount")
                )

            d.exchange_gain_loss = base_allocated_amount - allocated_amount_in_ref_exchange_rate

        return base_allocated_amount

    def set_unallocated_amount(self):
        self.unallocated_amount = 0
        if not self.party:
            return

        deductions_to_consider = sum(
            flt(d.amount) for d in self.get("deductions") if not d.is_exchange_gain_loss
        )
        included_taxes = self.get_included_taxes()

        if self.payment_type == "Receive" and self.base_total_allocated_amount < (
            self.base_paid_amount + deductions_to_consider
        ):
            self.unallocated_amount = (
                flt(
                    self.base_paid_amount + deductions_to_consider
                    - self.base_total_allocated_amount - included_taxes,
                    self.precision("unallocated_amount"),
                )
                / self.source_exchange_rate
            )
        elif self.payment_type == "Pay" and self.base_total_allocated_amount < (
            self.base_received_amount - deductions_to_consider
        ):
            self.unallocated_amount = (
                flt(
                    self.base_received_amount - deductions_to_consider
                    - self.base_total_allocated_amount - included_taxes,
                    self.precision("unallocated_amount"),
                )
                / self.target_exchange_rate
            )

        is_pra = self._is_payment_rate_allocation()
        if is_pra and self.unallocated_amount:
            payment_fc, rate = self._get_payment_fc_and_rate()
            total_allocated_fc = sum(
                flt(ref.allocated_amount) for ref in self.references
                if flt(ref.allocated_amount)
            )
            unallocated_fc = flt(
                payment_fc - total_allocated_fc, self.precision("paid_amount")
            )
            precision_threshold = 1 / (10 ** self.precision("paid_amount"))
            if abs(unallocated_fc) < precision_threshold:
                self.unallocated_amount = 0.0

    def _get_payment_fc_and_rate(self):
        if self.payment_type == "Pay":
            return flt(self.paid_amount), flt(self.source_exchange_rate)
        elif self.payment_type == "Receive":
            return flt(self.received_amount), flt(self.target_exchange_rate)
        return 0, 0
```

---

**Comment by Asobu01** [comment]
Thanks for the clarification on "@DipenFrappe you can show this code to frappe developer it solve the problem and make erpnext...".
I will align with these requirements and post a concrete checkpoint in ~20 minutes (repro/logs or patch summary), then submit a draft PR in ~8h.

**Comment by DipenFrappe** [comment]
@Manaa0-0, you are mixing 2 different issues

1st below setting is for users who want to create invoices in different currencies just for printing purposes and want to do accounting in the base currency. No exchange gain, loss booking, no multi-currency accounting.

If you want to do multicurrency, you should not enable the setting below and follow the multicurrency flow.

<img width="1276" height="756" alt="Image" src="https://github.com/user-attachments/assets/31bb2292-50d3-4837-a9e1-716a053d4f9a" />

the 2nd thing you asked is a known feature request in which you want to do a multi-currency accounting for different currencies against a single party/customer, which is currently not possible, and you need to create a separate party/customer for each currency.

**Comment by Manaa0-0** [comment]
@DipenFrappe thank the team fix the problem partially 712403aae4dd204913a2df0ef0fc5dfb7cf1d37f
